### ISSUE TITLE
Legacy XML reports should preserve new line characters in message attributes

### ISSUE THREAD

**Issue by gtach2o** [junit-team/junit5#4174]
Hello!

After switching test execution from using gradle with junit platform to console launcher I noticed that 
value of message attribute in JUnit xml report loosing new line characters.
e.g.
Using gradle xml :
```xml
<error message="line 1&#10; line 2&#10; line 3&#10;"...
```
Using Console launcher xml:
```xml
<error message="
line 1
line 2
line 3
"
```

## Steps to reproduce

1. Run tests using gradle with junit platform. Throw multiline exception in test.
2. Run tests using console launcher, throw multiline exception in test.

## Context

`org.junit.platform:junit-platform-console-standalone:1.11.3`
`org.junit.jupiter:junit-jupiter-engine:5.11.3`
`JRE 23`

As I understand console launcher uses new open/xml report generator and gradle platform launcher uses legacy report generator, right?

**Comment by marcphilipp** [comment]
@gtach2o I'm not sure what you mean. In your example, there are newline characters in the attribute value. While Gradle seems to replace `\n` using `&#10;`, the Console Launcher includes them as-is. This is well-formed XML. Does it cause any issues downstream?

**Comment by marcphilipp** [comment]
I've looked into it in #4180. However, the closest I can get with the current STaX-based infrastructure is this:

```xml
<failure message="line 1&amp;#10; line 2">...</failure>
```

That's because when passing `"line 1&#10; line 2"` to `javax.xml.stream.XMLStreamWriter#writeAttribute(String, String)` it will replace all `&` characters with `&amp;`.

@gtach2o Could you please check whether that kind of "double-escaping" works with your `XmlAttribute`-based processing?

**Comment by gtach2o** [comment]
> That's because when passing `"line 1&#10; line 2"` to `javax.xml.stream.XMLStreamWriter#writeAttribute(String, String)` it will replace all `&` characters with `&amp;`.

I think we only need to replace `\n` with `&#10;`. 
Thrown exception (error or failure) contains `\n` not `&#10;` after processing with `XMLStreamWriter` xml will have 
```xml
<failure message="line 1&#10; line 2&#10;">...</failure>
```

**Comment by gtach2o** [comment]
How gradle with
```groovy
test {
    useJUnitPlatform()
}
```
does it?

**Comment by gtach2o** [comment]
Maybe some double-escape avoiding mechanism can be applied 
```java
String escapeXmlAttribute(String input) {
        StringBuilder result = new StringBuilder();
        int length = input.length();
        for (int i = 0; i < length; i++) {
            char c = input.charAt(i);
            if (c == '&' && input.startsWith("&#10;", i)) {
               result.append("&#10;");
                i += 4;
            } else if (c == '\n') {
              result.append("&#10;");
            } else {                

result.append(StringEscapeUtils.escapeXml10(String.valueOf(c)));
            }
        }
        return result.toString();
}
```
or something like that, but to not reinvent the wheel would be better to look up how it s done during `useJUnitPlatform()`

**Comment by marcphilipp** [comment]
I did some more digging. The problem is that the default implementation of `XMLStreamWriter` does not replace `\n`, `\r`, and `\t` in attribute values while it does do so for `"`/`'` and `&`. I haven't found a way to customize that behavior.

However, adding [Woodstox](https://github.com/FasterXML/woodstox) to the classpath fixes the problem since it does escape those characters by default. It would even allow to configure a custom escaping strategy since it's [STaX2](https://github.com/FasterXML/stax2-api)-compatible but we don't need that.

I don't think we should add Woodstox as a dependency to junit-platform-reporting, though. Doing so could affect the code under test which would suddenly use a different XML parser/writer implementation. Thus, I've added test coverage and documentation in #4184.

@gtach2o Could you please verify that adding Woodstox to your test runtime classpath solves the problem for you as well?

**Comment by marcphilipp** [comment]
> How gradle with
> 
> test {
>     useJUnitPlatform()
> }
> 
> does it?

FWIW, Gradle uses [its own XML writer](https://github.com/gradle/gradle/blob/c8143c9bcf412a2d19053b8c926c7787c1dc8087/subprojects/core/src/main/java/org/gradle/internal/xml/SimpleXmlWriter.java).

**Comment by gtach2o** [comment]
> [@gtach2o](https://github.com/gtach2o) Could you please verify that adding Woodstox to your test runtime classpath solves the problem for you as well?

There is no way to do it without adding third party dependency ?

**Comment by gtach2o** [comment]
Can you create a custom wrapper class around XMLStreamWriter that overrides the writeAttribute method to handle escaping these characters? The rest of methods can use the original writer.

```java

public class CustomXMLStreamWriter implements XMLStreamWriter {
    private final XMLStreamWriter delegate;

    public CustomXMLStreamWriter(XMLStreamWriter delegate) {
        this.delegate = delegate;
}

@Override
    public void writeAttribute(String localName, String value) throws XMLStreamException {
       
        String escapedValue = value
            .replace("\n", "&#10;")
            .replace("\r", "&#13;")
            .replace("\t", "&#9;");
        delegate.writeAttribute(localName, escapedValue);
}
. . .
```
Then use it in legacy.xml.XmlReportWriter
```java
      XMLStreamWriter xmlWriter = new CustomXMLStreamWriter(factory.createXMLStreamWriter(out)); 
```

Maybe this and no need in woodstox?

**Comment by marcphilipp** [comment]
No, that wouldn't help since the delegate would then escape `&` so we'd end up with `&amp;10;`.

But maybe I can wrap the underlying `Writer` and replace them there. I'll give that a try.

**Comment by marcphilipp** [comment]
Seems to work:

- #4186

**Comment by gtach2o** [comment]
> Seems to work:
> 
>     * [Escape whitespace chars in XML attribute values #4186](https://github.com/junit-team/junit5/pull/4186)

Cool!
So, woodstox is there only for tests?
Are there jar files somewhere I can download to try it out?

**Comment by marcphilipp** [comment]
Yes, Woodstox is only used for testing (as an additional test task).

I've now merged the PR and you can consume 5.12.0-SNAPSHOT versions from https://oss.sonatype.org/content/repositories/snapshots/

**Comment by marcphilipp** [comment]
We will include this in 5.11.4 which will be released toward the end of this week.
